psi4/src/psi4/libmints/thc_eri.cc

```diff
@@ -0,0 +1,354 @@
+    #pragma omp atomic
+                        (*E_PQ)(p, q) += prefactor1 * (*E_temp)(q, dm * nn + dn) * (*x1_)(p, m) * (*x1_)(p, n);
+                    } // end n
+                } // end m
+            } // end q
+        } // end p
+    } // end MN
+
+    return E_PQ;
+}
+
+SharedMatrix LS_THC_Computer::build_E_df() {
+    /**
+     * Parrish et al. 2012 Procedure 3
+    */
+    
+    size_t nbf = primary_->nbf();
+    size_t naux = auxiliary_->nbf();
+    size_t rank = x1_->nrow();
+
+    int nthreads = 1;
+#ifdef _OPENMP
+    nthreads = Process::environment.get_n_threads();
+#endif
+
+    auto zero = BasisSet::zero_ao_basis_set();
+    IntegralFactory factory(auxiliary_, zero, primary_, primary_);
+    std::vector<std::shared_ptr<TwoBodyAOInt>> eri_computers(nthreads);
+
+    eri_computers[0] = std::shared_ptr<TwoBodyAOInt>(factory.eri());
+#pragma omp parallel for
+    for (int thread = 1; thread < nthreads; ++thread) {
+        eri_computers[thread] = std::shared_ptr<TwoBodyAOInt>(eri_computers[0]->clone());
+    }
+
+    size_t nshell_aux = auxiliary_->nshell();
+    size_t nshellpair = eri_computers[0]->shell_pairs().size();
+    size_t nshelltriplet = nshell_aux * nshellpair;
+
+    SharedMatrix E_temp = std::make_shared<Matrix>(rank, naux);
+
+#pragma omp parallel for
+    for (size_t MNP = 0; MNP < nshelltriplet; ++MNP) {
+
+        size_t MN = MNP % nshellpair;
+        size_t P = MNP / nshellpair;
+        int thread = 0;
+#ifdef _OPENMP
+        thread = omp_get_thread_num();
+#endif
+        auto bra = eri_computers[thread]->shell_pairs()[MN];
+        size_t M = bra.first;
+        size_t N = bra.second;
+
+        int np = auxiliary_->shell(P).nfunction();
+        int pstart = auxiliary_->shell(P).function_index();
+        int nm = primary_->shell(M).nfunction();
+        int mstart = primary_->shell(M).function_index();
+        int nn = primary_->shell(N).nfunction();
+        int nstart = primary_->shell(N).function_index();
+
+        // TODO: Implement screening to make this more efficient
+        eri_computers[thread]->compute_shell(P, 0, M, N);
+        const auto &buffer = eri_computers[thread]->buffers()[0];
+
+        double prefactor = (M == N) ? 1.0 : 2.0;
+
+        for (size_t r = 0; r < rank; ++r) {
+            for (size_t p = pstart, index = 0; p < pstart + np; ++p) {
+                for (size_t m = mstart; m < mstart + nm; ++m) {
+                    for (size_t n = nstart; n < nstart + nn; ++n, ++index) {
+#pragma omp atomic
+                        (*E_temp)(r, p) += prefactor * buffer[index] * (*x1_)(r, m) * (*x1_)(r, n);
+                    } // end n
+                } // end m
+            } // end p
+        } // end r
+    } // end MNP
+
+    FittingMetric J_metric_obj(auxiliary_, true);
+    J_metric_obj.form_fitting_metric();
+    auto J_metric = J_metric_obj.get_metric();
+    
+    int nremoved = 0;
+    auto Jinv = J_metric->pseudoinverse(1.0e-14, nremoved);
```

Comment:
Perhaps it would be a good idea to leave a comment here in the code about why 1.0e-14 is an appropriate threshold? What would happen if this was set lower or higher?
Also, maybe it would be a good idea to add a check here that prints a warning and/or throws if `nremoved` is too high for whatever reason. Maybe consider getting the print level from `options.get_int("PRINT")`, and always print `nremoved` at higher levels.

Reply:
1.0e-14 is around machine precision for a double on most OS. Anything lower would lead to numerical instabilities, and anything higher would remove too many singular values. There is precedent in Psi for this value

https://github.com/psi4/psi4/blob/master/psi4/include/psi4/psi4-dec.h#L52

Reply:
It's good practice to use `PSI_ZERO` directly, rather than hard-coding more numbers.

Reply:
Yep, changed it to use `PSI_ZERO` instead of hard-coded `1.0e-14`